include/quic/datagram.hpp

```diff
@@ -1,16 +1,71 @@
 #pragma once
 
+#include "udp.hpp"
 #include "utils.hpp"
 
 namespace oxen::quic
 {
+    extern std::atomic<bool> enable_test_features;
+    extern std::atomic<int> test_counter;
+
     class IOChannel;
     class Connection;
     class Endpoint;
     class Stream;
 
     // IO callbacks
-    using dgram_data_callback = std::function<void(bstring_view)>;
+    using dgram_data_callback = std::function<void(bstring)>;
+
+    using dgram_buffer = std::deque<std::pair<uint16_t, std::pair<bstring_view, std::shared_ptr<void>>>>;
+
+    struct prepared_datagram
+    {
+        uint64_t id;                  // internal ID for ngtcp2
+        std::array<uint8_t, 2> dgid;  // optional transmitted ID buffer (for packet splitting)
+        std::array<ngtcp2_vec, 2> bufs;
+        size_t bufs_len;  // either 1 or 2 depending on how much of data is populated
+
+        const ngtcp2_vec* data() const { return bufs.data(); }
+        size_t size() const { return bufs_len; }
+    };
+
+    struct received_datagram
+    {
+        uint16_t id{0};
+        bstring data{};
+        // -1 = first half, 1 = second half
+        int part{0};
+
+        received_datagram() = default;
+        explicit received_datagram(uint16_t dgid, bstring_view d) : id{dgid}, data{d} { part = (dgid % 4 == 2) ? -1 : 1; };
```

Comment:
We know that *at most* the assembled datagram will be `d.size() + MAX_DATAGRAM_SIZE` (this is probably the wrong name for that constant), so we can reserve it with:
```suggestion
        explicit received_datagram(uint16_t dgid, bstring_view d) : id{dgid}, part{(dgid % 4 == 2) ? -1 : 1} {
            data.reserve(d.size() + MAX_DATAGRAM_SIZE);
            data.append(d);
        }
```
and then we fill out this buffer with the remaining packet when it arrives without having to copy and allocate a new buffer.

Reply:
awesome